include/javabind/collection.hpp

```diff
@@ -289,11 +289,11 @@ namespace javabind
     /**
      * Converts a native set (e.g. a [set] or [unordered_set]) into a Java Set.
      */
-    template <template<typename...> typename S, typename T>
-    struct JavaSetType : AssignableJavaType<S<T>>
+    template <typename T, template<typename...> typename S>
+    struct JavaSetType : AssignableJavaType<T>
     {
-        using native_type = S<T>;
-        using native_boxed_type = S<boxed_t<T>>;
+        using native_type = T;
+        using native_boxed_type = S<boxed_t<typename T::key_type>>;
```

Comment:
I think we may need to propagate `Compare` (for ordered), `Hash` and `KeyEqual` (for unordered), and `Allocator`. We might also need to generate wrapper functions e.g. to adapt a hash function on `key_type` to a hash on `boxed_t<key_type>`.

Reply:
The additional template parameters are propagated here as well now.

We don't need a hash adapter because the native_boxed_type is only used to determine the java counterpart. In the native code we still use native_type to access the c++ object.

Reply:
I added tests where an int is used as value / key with a custom hash